### PR TITLE
Tests implementation for SYCL2020 `accessor` constructors

### PR DESCRIPTION
**Pull request by andreyromanof**
This PR provides tests for `sycl::accessor` constructors. Tests cover only non-deprecated access modes and targets.

Depends on:
- #321 
- #322 
- #324

### CODE REVIEW

**Comment by andreyromanof** [comment]
Have to add to filter hipsycl and computecpp as they doesn't support marray([hipsycl](https://github.com/KhronosGroup/SYCL-CTS/runs/5830730001?check_suite_focus=true#step:8:108) error message; [computecpp](https://github.com/KhronosGroup/SYCL-CTS/runs/5830729872?check_suite_focus=true#step:8:15) error message).

File: ci/computecpp.filter
Comment (bader) [inline ci/computecpp.filter:1]: Can we disable the test cases instead of the whole test?

File: tests/accessor/accessor_common.h
Comment (keryell) [inline tests/accessor/accessor_common.h:185]: Use early exit
```suggestion
  for (size_t i = 0; i < N; ++i)
    if (left[i] != right)
      return false;
  return true;
```

File: tests/accessor/accessor_common.h
Comment (keryell) [inline tests/accessor/accessor_common.h:198]: Idem

File: tests/accessor/generic_accessor_constructors.hpp
Comment (keryell) [inline tests/accessor/generic_accessor_constructors.hpp:29]:
```suggestion
    auto get_acc_functor = [] {
```

File: tests/accessor/accessor_common.h
Comment (keryell) [reply tests/accessor/accessor_common.h:198]: More generally I have the feeling that all these double definitions for `sycl::vec` and `sycl::marray` could be factorized.
Perhaps by adding the equivalent of a C++20 concept true for `sycl::vec` or `sycl::marray`, but implemented by `std::enable_if_t`.

**Review by keryell — CHANGES_REQUESTED** [review]
Great contribution.

File: ci/computecpp.filter
Comment (andreyromanof) [reply ci/computecpp.filter]: I added `#ifdef` command to disable part of code and `DISABLE_FOR_TEST_CASE` to disable some tests . This should allow us to remove whole test from the filter.

File: tests/accessor/accessor_common.h
Comment (andreyromanof) [reply tests/accessor/accessor_common.h]: Fixed in #324

File: tests/accessor/accessor_common.h
Comment (andreyromanof) [reply tests/accessor/accessor_common.h]: As it's yet another one implementation of `value_helper` functions ([first impl](https://github.com/KhronosGroup/SYCL-CTS/blob/b00bde0e3580d0c8a3cd0c7f7c46f60c5cb4d596/tests/extension/oneapi_device_global/device_global_common.h#L84) in `device_global` tests), I moved these functions in separate file in PR #324

File: tests/accessor/generic_accessor_constructors.hpp
Comment (andreyromanof) [reply tests/accessor/generic_accessor_constructors.hpp]: Fixed

**Review by keryell — APPROVED** [review]
Thanks.

**Comment by keryell** [comment]
@illuhad @ProGTX any second review on this big PR with a lot of other ones depending on it?

File: tests/accessor/accessor_common.h
Comment (psalz) [inline tests/accessor/accessor_common.h:17]: I think we usually use `<angle brackets>` for third-party includes!

**Review by psalz — CHANGES_REQUESTED** [review]
Thanks! While the fact that none of the implementations can currently compile these tests is obviously not great, I think we agreed that such PRs can still be merged to support test-driven development. Do you think this situation could be at least somewhat remedied if we had a `DISABLED_FOR_SECTION`?

File: tests/accessor/accessor_common.h
Comment (yuriykoch) [reply tests/accessor/accessor_common.h:17]: Actually, looks like it would be better to use the following:
```
#include "../common/common.h"
// instead of <catch2/catch_test_macros.hpp>
```
and
```
#include <catch2/catch_template_test_macros>
#include "../common/macros.h"
// instead of <catch2/catch_template_test_macros.hpp>
```
because of `#undef FAIL` within [../common/macros.h](https://github.com/KhronosGroup/SYCL-CTS/blob/0a5d5d689124edcc2280a0d7d36d2e16513904f5/tests/common/macros.h#L16)
@psalz, @andreyromanof   what do you think?

Comment (psalz) [reply tests/accessor/accessor_common.h:17]: You're right, including `<catch2/catch_test_macros.hpp>` is superfluous in this case, just including "`../common/common.h"` should be sufficient (as that transitively includes `macros.h`). Also the custom `FAIL` macro is actually not really needed in this case, it only exists for backwards compatibility with the old `FAIL` macro that received a `logger`.

File: tests/accessor/accessor_common.h
Comment (andreyromanof) [reply tests/accessor/accessor_common.h:237]: I'm reusing this common function for `local_accessor` tests. As I understand, we can't use `local_accessor` in `single_task` or `parallel_for` with `sycl::range` as argument, so I use `parallel_for_work_group` instead for compability.

>A local accessor must not be used in a [SYCL kernel function](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sycl-kernel-function) that is invoked via single_task or via the simple form of parallel_for that takes a range parameter. [4.7.6.11. Local accessor](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:accessor.local)

File: tests/accessor/accessor_common.h
Comment (andreyromanof) [reply tests/accessor/accessor_common.h]: Removed extra includes and added angle brackets to catch header `catch2/matchers/catch_matchers.hpp` include

File: tests/accessor/accessor_common.h
Comment (yuriykoch) [reply tests/accessor/accessor_common.h:379]: Upd: found `AccType` dependency in #325 : https://github.com/KhronosGroup/SYCL-CTS/pull/325/files#diff-8067235ffc00be4a68e8da8448acbf789b606099c5dbc1e969df066646bd2f64R410.

File: tests/accessor/accessor_common.h
Comment (psalz) [reply tests/accessor/accessor_common.h:236]: Ah yes, makes sense!

File: tests/accessor/accessor_common.h
Comment (andreyromanof) [reply tests/accessor/accessor_common.h]: Fixed

File: tests/accessor/accessor_common.h
Comment (yuriykoch) [reply tests/accessor/accessor_common.h:379]: > For me, the link goes just to the Files changed tab not to a concrete place.

Yep, looks like github restriction. It doesn't show exact line in `accessor_common.h` until the diff is loaded by clicking on "load diff".

> I stood for waiting until all PRs are merged, not only this one.

Makes sense. Unfortunatelly looks like I'll be a bit blocked until then, but I'll try to find a workaround.

> There is a difference in API. Zero dimension accessors behave like a scalar variable without a subscript operator, unlike 1-2-3 dimension accessors([spec](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_interface_for_buffer_command_accessors) lines 155 and 158). That's why read_write_zero_dim_acc functions exist, as they change accessor without invoking the subscript operator. No doubt we can constantly check if we working with a zero dimension accessor, to separate usage of 1-2-3 dimension accessor from 0 dimension accessor, but IMO it will be too many constexpr checks, that can be avoided by a different function.

Look, I totally agree that havin a separate function within `accessor_common.h` might be really nice.
But as I see it, there is no reason to present a separate API to the user of `accessor_common.h`, as it would require to use `if constexpr (Dimension == 0)` even if there is no difference at all from the point of the caller. Looks like even `sycl::range` usage might be resolved in a generic way.
Still, moving refactoring out of this PR means this question is out of scope of this PR as well.

**Comment by keryell** [comment]
So, what do we do with this?
Do we merge this as is and go on improving with other PR?
@psalz you asked for some changes. Are you happy now?

**Comment by andreyromanof** [comment]
> Do we merge this as is and go on improving with other PR?

I think it's a right decision to make a improving PR after merging current PRs with accessor tests. It will allow us to make changes in one common branch.

**Comment by andreyromanof** [comment]
> Do you think this situation could be at least somewhat remedied if we had a DISABLED_FOR_SECTION?

Sorry for delay answer. In other cases it can help, but in accessors tests I think it won't make a big deal. For example DPCPP doesn't implement `sycl::target` that used in all sections, so if we theoretically could disable sections, we would disable all of them for DPCPP.

**Review by psalz — APPROVED** [review]

**Pull request merged**
